### ISSUE TITLE
Planned Cross-Platform support?

### ISSUE THREAD

**Issue by codynhat** [izqui/Taylor#13]
With Swift being released as open source soon, are there any plans to make this project cross-platform? Specifically, Linux? I know Apple is planning on releasing a compiler for Linux. It would be awesome to have a robust Swift web framework that can run on a Linux web server.

**Comment by Danappelxx** [comment]
+1

**Comment by izqui** [comment]
Yes! That's the idea, but we it is difficult to anticipate without knowing what will be allowed in Linux Swift.

The first step I took towards that goal is abstracting the [Socket protocol](https://github.com/izqui/Taylor/blob/master/Taylor/Socket.swift), in case we can't use CocoaAsyncSocket when it launches, to make the transition easily.

The other thing that bugs me is that it relies heavily on Foundation (lots of NSData around) and it is unclear whether it could be used from Linux.

Long story short, making it cross-platform is definitely the goal and we would love to have as much help as posible.

Thanks for the interest.

**Comment by codynhat** [comment]
Looks like Node.js is built on https://github.com/libuv/libuv. Could that be an option?

**Comment by Danappelxx** [comment]
There's a lot of socket libraries around in c and objective-c - I even found [one](https://github.com/AlwaysRightInstitute/SwiftSockets) in Swift without any Foundation dependencies. I agree with @izqui though, you can't really predict how much control over Swift Apple is going to give, so it's kind of useless to find another dependency that could be just as likely to be unusable when Swift does go open source. I would instead try to get rid of as much Foundation use in Taylor as possible, since it is very unlikely that Apple will open source it.

**Comment by perezpaya** [comment]
@Danappelxx +1

**Comment by izqui** [comment]
I started implementing the socket protocol with the library @Danappelxx introduced.

I got it to work, but I have this little issue that I'm trying to figure out now https://github.com/AlwaysRightInstitute/SwiftSockets/issues/10

I started working on it in the `swiftsockets` branch https://github.com/izqui/Taylor/blob/swiftsockets/Taylor/Socket.swift

**Comment by izqui** [comment]
Ups, something is weird with the headers also, will have to look at it in more detail.

**Comment by izqui** [comment]
Solved! `swiftsockets` is working now :)

**Comment by Danappelxx** [comment]
Nice!

**Comment by Danappelxx** [comment]
Now that Swift is open source, I think we should work on making Taylor linux compatible. I put some thought into this and I believe the biggest thing stopping us is the use of Foundation - most importantly NSData. The solution I see popping up is using `[Int8]` for bytes, and I think we should do the same.

**Comment by jhoughjr** [comment]
Might be old news, but I believe Foundation is working in Linux under Swift 2.2 according to Swift.org.
